src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/HttpConnectionPool.Http3.cs

```diff
@@ -88,7 +88,16 @@ internal sealed partial class HttpConnectionPool
 
                 if (!TryGetPooledHttp3Connection(request, out Http3Connection? connection, out HttpConnectionWaiter<Http3Connection?>? http3ConnectionWaiter))
                 {
-                    connection = await http3ConnectionWaiter.WaitWithCancellationAsync(cancellationToken).ConfigureAwait(false);
+                    using Activity? waitForConnectionActivity = ConnectionSetupDiagnostics.StartWaitForConnectionActivity(authority);
```

Comment:
Note that we call `StopConnectionSetupActivity()` at L274. It does more than just stopping the activity (not prepared for double-stopping) and I prefer to do the stop on the success path before the `InitQuicConnection() -> MarkConnetionAsEstablished()` calls in order to be consistent with H/1.1 H/2. I wanted to avoid defining a special method in `ConnectionSetupDiagnostics` that would check for `Activity.IsStopped` just for the H3 case.

I guess this logic deserves more comments?

Reply:
> the start and stop of this activity should closely follow queueStartingTimestamp.

Implemented.

The problem with the existing logic is that we will emit a very short `wait_for_connection` span and near-zero `http.client.request.time_in_queue` duration metric when there is a connection available in the pool. This is not something I would touch in this PR, but maybe we can improve this in long term?

Reply:
Agreed, we already keep stream count in H3Connection, we could use it to know if we're gonna wait or not. But in general, it would be good to make the stream opening as close as possible to H2 to avoid these quirks. But that's 10.0 issue. I'll file it.

Reply:
Filed: https://github.com/dotnet/runtime/issues/104783